### PR TITLE
Feature/wind sensor

### PR DESCRIPTION
**Pull request by Boto1**

### CODE REVIEW

File: generator/main.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+func setInterval(function func(), seconds int) {
+	ticker := time.NewTicker(time.Duration(seconds) * time.Second)
+	done := make(chan bool)
+
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				function()
+			case <-done:
```
Comment (polluxx) [inline generator/main.go]: if it's not used - remove the done channel, it's wasting memory resources

File: generator/main.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+	}()
+}
+
+func main() {
+	setInterval(GenerateWindEvents, 5)
+	http.ListenAndServe(":1234", nil)
+}
+
+func GenerateWindEvents() {
+	if checkForStatus() != 1 {
```
Comment (polluxx) [inline generator/main.go]: Manage error here, not only status, you should be able to see what do you get from your response and log it at least

File: generator/main.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+	if mean == -1 {
+		return
+	}
+	var windPower float64 = math.Round(normPower(mean)*10) / 10
+	var beaufortValue uint8 = uint8(math.Round(windPower / 10))
+	if beaufortValue == 0 {
+		beaufortValue = 1
+	}
+
+	reqBody, err := json.Marshal(map[string]interface{}{
```
Comment (polluxx) [inline generator/main.go]: Use struct for request payload.

File: generator/main.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+	if err != nil {
+		fmt.Println("Could not convert to json")
+	}
+
+	req, err := http.Post("http://localhost:8080/wind/event", "application/json", bytes.NewBuffer(reqBody))
+
+	if err != nil {
+		fmt.Println("Cannot create wind event", err)
+	} else {
```
Comment (polluxx) [inline generator/main.go]: remove else, just return an empty value after the print

File: generator/main.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+		fmt.Println("Wind event created")
+		defer req.Body.Close()
```
Comment (polluxx) [inline generator/main.go]: If it's the last line - defer is unneeded; `defer` means that we will call this function when the main scope function (in which it was called) will finish

File: generator/main.go [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+		return -1
+	}
+
+	data, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return -1
+	}
+
+	var power float64
+	err = json.Unmarshal(data, &power)
```
Comment (polluxx) [inline generator/main.go]: What is `data`? If it's a string - just convert the string to float64, reflection, in this case, is a bad practise

File: server/domain/devices/postgre.go [hunk 7]
```diff
@@ -17,12 +18,13 @@ const (
...
 
 var NOT_FOUND = errors.New("record not found")
+var BAD_STATUS = errors.New("status is not correct")
 
 //type Storage *gorm.DB
 var Storage *gorm.DB
 
 func Connect() (*gorm.DB, error) {
-	db, err := gorm.Open("postgres", fmt.Sprintf("host=%s port=%s user=%s dbname=%s password=%s",
+	db, err := gorm.Open("postgres", fmt.Sprintf("host=%s port=%s user=%s dbname=%s password=%s sslmode=disable",
```
Comment (polluxx) [inline server/domain/devices/postgre.go:31]: is that a particular issue with windows? Why is ssl connection here?

File: server/domain/devices/wind.go [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+func (t *Wind) FindManyEvents(from string, to string) ([]WindEvent, error) {
+	var events []WindEvent
+
+	var err error
+	query := Storage.Where("created BETWEEN ? AND ?", from, to)
+	query = query.Select("date_trunc('minute', created) as created, round(avg(power), 1) as power, degrees(atan(sum(sin(radians(direction))) / sum(cos(radians(direction))))) as direction, min(event_id) as event_id, round(avg(beaufort_value), 0) as beaufort_value")
+	err = query.Group("date_trunc('minute', created)").Order("min(event_id)").Find(&events).Error
+	if err != nil {
+		// returning custom DB error message
+		err = NOT_FOUND
```
Comment (polluxx) [inline server/domain/devices/wind.go]: It may be not only `not found` error here.
Better to remove the error mutation and return it as is in all DB methods.
The masking error as added as the example.

File: server/domain/devices/wind.go [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+	if err != nil {
+		fmt.Printf("Wind Sensor is not created")
+		return StatusFailure, NOT_FOUND
+	}
+
+	sensor.Status = status
+	return status, Storage.Save(&sensor).Error
+}
+
+func (t *Wind) FindManyEvents(from string, to string) ([]WindEvent, error) {
```
Comment (polluxx) [inline server/domain/devices/wind.go]: If there are many events - return a pointer

File: server/domain/routers/dashboard.go [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+	sensors, err := sensor.FindManySensors()
+	// testing custom error response
+	if err == devices.NOT_FOUND {
+		http.Error(w, errors.New("no sensors in the database").Error(), http.StatusNotFound)
+		return
+	}
+
+	response, err := json.Marshal(sensors)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusNotFound)
```
Comment (polluxx) [inline server/domain/routers/dashboard.go]: Better to send internal server error status in code

File: server/domain/routers/wind.go [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+	device, err := wind.Get()
+	// testing custom error response
+	if err == devices.NOT_FOUND {
+		http.Error(w, errors.New("the device is not found").Error(), http.StatusNotFound)
+		return
+	}
+
+	response, err := json.Marshal(device)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusNotFound)
```
Comment (polluxx) [inline server/domain/routers/wind.go]: same with status here

File: server/domain/routers/wind.go [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+	w.Write(response)
+}
+
+func UpdateWindSensor(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+	defer r.Body.Close()
+
+	wind := devices.Wind{}
+	_, err := wind.Get()
+	if err == devices.NOT_FOUND {
+		http.Error(w, errors.New("the device is not found").Error(), http.StatusNotFound)
```
Comment (polluxx) [inline server/domain/routers/wind.go]: Just add a plain text here, no need to make an error and then get the text from this error

File: server/domain/routers/wind.go [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+		return
+	}
+
+	type Data struct {
```
Comment (polluxx) [inline server/domain/routers/wind.go]: Structs should be defined at the top of the package

File: server/domain/routers/wind.go [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+		Status devices.SensorState
+	}
+	var data Data
+
+	rBytes, err := ioutil.ReadAll(r.Body)
+	convErr := json.Unmarshal(rBytes, &data)
+	if convErr != nil || err != nil {
```
Comment (polluxx) [inline server/domain/routers/wind.go]: Wrong, if you got an error the rBytes will be nil, which will cause panic.

File: server/domain/routers/wind.go [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+	windEvents, err := wind.FindManyEvents(from, to)
+
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	response, err := json.Marshal(windEvents)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusNotFound)
```
Comment (polluxx) [inline server/domain/routers/wind.go]: fix status please

File: server/domain/routers/wind.go [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+func GetLastDate(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+	wind := devices.Wind{}
+	windEvent, err := wind.FindOneEvent()
+
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	response, err := json.Marshal(windEvent.Power)
```
Comment (polluxx) [inline server/domain/routers/wind.go]: No need to marshal non-structured value

File: server/domain/routers/wind_test.go [hunk 17]
```diff
@@ -0,0 +1,97 @@
+package routers_test
+
```
Comment (polluxx) [inline server/domain/routers/wind_test.go]: Remove test if it's not testing your logic

File: server/main.go [hunk 18]
```diff
@@ -27,20 +29,32 @@ func main() {
...
 		return
 	}
 
+	err = s.Devices.Wind.CreateSensor()
+	if err != nil {
+		log.Fatal(fmt.Printf("Error creating device: %v \n", err))
+		return
+	}
+
+	c := cors.New(cors.Options{
```
Comment (polluxx) [inline server/main.go:66]: Then remove `headers setting` in your handlers if it uses wildcard and methods here

File: server/main.go [hunk 19]
```diff
@@ -27,20 +29,32 @@ func main() {
...
+		AllowedOrigins: []string{"*"},
+		AllowedMethods: []string{"GET", "POST", "DELETE", "PUT", "OPTIONS"},
+	})
+
 	// init server
-	log.Fatal(http.ListenAndServe(":8080", router))
+	log.Fatal(http.ListenAndServe(":8080", c.Handler(router)))
 }
```
Comment (polluxx) [inline server/main.go:73]: Please also remove binaries from your PR

**Review by polluxx — CHANGES_REQUESTED** [review]

File: server/domain/routers/wind.go (hunk 16 above)
Comment (Boto1) [reply server/domain/routers/wind.go]: `var response string = fmt.Sprintf("%f", windEvent.Power)
w.Write([]byte(response))` is it okay to do like this?

File: server/domain/routers/wind.go (hunk 13 above)
Comment (Boto1) [reply server/domain/routers/wind.go]: okay, but what should i do if i dont wanna export a struct?

File: server/domain/routers/dashboard.go (hunk 10 above)
Comment (Boto1) [reply server/domain/routers/dashboard.go]: i cant remember that i wrote this file, anyway ok:)

File: server/domain/devices/wind.go (hunk 9 above)
Comment (Boto1) [reply server/domain/devices/wind.go]: slice is already a pointer like in cpp, isnt it?

File: server/domain/devices/postgre.go (hunk 7 above)
Comment (Boto1) [reply server/domain/devices/postgre.go:31]: yeah, i googled this issue, for some reason i cant connect to the db without disabling ssl

File: generator/main.go (hunk 6 above)
Comment (Boto1) [reply generator/main.go]: not sure i got u right, so `readall()` returns slice of bytes, i cant just convert it directly to float64..

File: client/package.json [hunk 20]
```diff
@@ -13,20 +13,31 @@
     "not ie < 9"
   ],
   "dependencies": {
+    "axios": "^0.19.0",
```
Comment (undead404) [inline client/package.json]: If the dependency is unstable, use exact version notation (`"0.19.0"` this time)

File: client/package.json [hunk 21]
```diff
@@ -13,20 +13,31 @@
...
     "body-parser": "~1.19.0",
+    "bootstrap": "~4.3.1",
+    "chart.js": "^2.8.0",
```
Comment (undead404) [inline client/package.json]: Use tilde notation on all stable dependencies please

File: server/domain/routers/wind.go (hunk 16 above)
Comment (polluxx) [reply server/domain/routers/wind.go]: yes

File: server/domain/devices/wind.go [hunk 22]
```diff
@@ -0,0 +1,106 @@
...
+	var events []WindEvent
+
+	var err error
+	query := Storage.Where("created BETWEEN ? AND ?", from, to)
+	query = query.Select("date_trunc('minute', created) as created, round(avg(power), 1) as power, degrees(atan(sum(sin(radians(direction))) / sum(cos(radians(direction))))) as direction, min(event_id) as event_id, round(avg(beaufort_value), 0) as beaufort_value")
```
Comment (polluxx) [inline server/domain/devices/wind.go:59]: Interesting formula, what is the math equivalent for that?

Comment (Boto1) [reply server/domain/devices/wind.go]: `(sin(15) + sin(10)) / (cos(15) + cos(10))` - its tangent of the angle, then i apply `atan()` func to find out the angle

Comment (Boto1) [reply server/domain/devices/wind.go:59]: `(sin(15) + sin(10)) / (cos(15) + cos(10))` - its the tangent of the angle, than i apply `atan()` func to find out this angle.

File: client/src/actions/wind-sensor.js [hunk 23]
```diff
@@ -0,0 +1,93 @@
...
+};
+
+export const changeWindStatus = status => async dispatch => {
+  dispatch(loadWindStatus());
+  try {
+    const { data } = await axios.put(`${apiURL}/wind`, {
+      status: status ? 1 : 0,
+    });
+    const delay = 1000;
+    setTimeout(() => dispatch(updateWindStatus(data)), delay);
```
Comment (undead404) [inline client/src/actions/wind-sensor.js:66]: What is this delay for?

File: client/src/actions/wind-sensor.js [hunk 24]
```diff
@@ -0,0 +1,93 @@
...
+  } catch (error) {
+    dispatch(windSensorFailure(error.message));
+  }
+};
+
+export const getWindSensorData = () => async (dispatch, getState) => {
+  const { windSensor } = getState();
+  if (!windSensor.info.Name) dispatch(windSensorLoading());
```
Comment (undead404) [inline client/src/actions/wind-sensor.js]: Please use lodash's `get` or destructuring here

File: client/src/actions/wind-sensor.js [hunk 25]
```diff
@@ -0,0 +1,93 @@
...
+  const queries = [
+    axios.get(`${apiURL}/wind/events`, {
+      params: {
+        from: windSensor.filterOption.from,
```
Comment (undead404) [inline client/src/actions/wind-sensor.js]: and also here

File: client/src/routes/wind/LogRecord.js [hunk 26]
```diff
@@ -0,0 +1,80 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import withStyles from 'isomorphic-style-loader/withStyles';
+import s from './css/Record.css';
```
Comment (undead404) [inline client/src/routes/wind/LogRecord.js]: Why do you store css files separately? Store all files related to one component together

File: client/src/routes/wind/LogRecord.js [hunk 27]
```diff
@@ -0,0 +1,80 @@
...
+          </defs>
+        </svg>
+        <svg
+          width={(rectWidth + rectGap) * maxRects - rectGap}
+          height={rectHeight}
+          viewBox={`0 0 ${(rectWidth + rectGap) * maxRects -
+            rectGap} ${rectHeight}`}
+        >
+          <clipPath id="clip-rects">
+            {Array.from(new Array(rects), (_, i) => (
```
Comment (undead404) [inline client/src/routes/wind/LogRecord.js]: lodash got `range` for this :-)

File: client/src/routes/wind/Logger.js [hunk 28]
```diff
@@ -0,0 +1,56 @@
+import React from 'react';
+import withStyles from 'isomorphic-style-loader/withStyles';
+import PropTypes from 'prop-types';
+import { connect } from 'react-redux';
+import Fade from 'react-reveal/Fade';
+import LogRecord from './LogRecord';
+import groupifyEvents, { selectOnlyTodayEvents } from '../../utils/logger';
+import s from './css/Logger.css';
+
+// eslint-disable-next-line no-magic-numbers
```
Comment (undead404) [inline client/src/routes/wind/Logger.js]: please remove this eslint directive, assign 450 to a constant giving it a name

File: client/src/routes/wind/RightPanel.js [hunk 29]
```diff
@@ -0,0 +1,159 @@
...
+    ).isRequired,
+    name: PropTypes.string.isRequired,
+  };
+
+  render() {
+    const { name, events } = this.props;
+    if (events.length === 0) {
+      return <WindStatus name={name} />;
+    }
+    // eslint-disable-next-line no-magic-numbers
```
Comment (undead404) [inline client/src/routes/wind/RightPanel.js]: Pls add `-1` to allowed magic numbers instead (in eslint config)

File: client/src/routes/wind/WindChart.js [hunk 30]
```diff
@@ -0,0 +1,148 @@
+/* eslint-disable react/jsx-handler-names */
```
Comment (undead404) [inline client/src/routes/wind/WindChart.js]: please remove all eslint disable directives, all these rules were added by intention, and correct your code accordingly

File: client/src/utils/logger.js [hunk 31]
```diff
@@ -0,0 +1,81 @@
+/* eslint-disable no-magic-numbers */
+/* eslint-disable no-param-reassign */
```
Comment (undead404) [inline client/src/utils/logger.js]: Please no

File: client/src/utils/wind-right-panel.js [hunk 32]
```diff
@@ -0,0 +1,81 @@
+/* eslint-disable no-magic-numbers */
```
Comment (undead404) [inline client/src/utils/wind-right-panel.js:1]: Approved

File: client/yarn.lock [hunk 33]
```diff
@@ -1,9747 +0,0 @@
-# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (undead404) [inline client/yarn.lock]: Why did you remove this?

**Review by undead404 — CHANGES_REQUESTED** [review]

File: client/src/actions/wind-sensor.js (hunk 23 above)
Comment (Boto1) [reply client/src/actions/wind-sensor.js:66]: for visual consistency, usually server responds quickly for updating sensor status and loader is blinking on the page, so i decided to add 1s delay thus it looks more natural for the user

File: client/src/routes/wind/LogRecord.js (hunk 26 above)
Comment (Boto1) [reply client/src/routes/wind/LogRecord.js]: to decrease the level of mess in files structure, as i have quite a few components and each of them has its own css file

File: client/src/routes/wind/LogRecord.js (hunk 27 above)
Comment (Boto1) [reply client/src/routes/wind/LogRecord.js]: yeah, but im not sure why do i need this for only one feature, i think my way of doing this is not so bad and unclear to replace it with a `range`

File: .gitignore [hunk 34]
```diff
@@ -28,4 +28,5 @@ yarn-error.log*
 # Misc
 .DS_Store
 ./server/go.sum
-server/server
```
Comment (undead404) [inline .gitignore:32]: Please restore this line as linux binaries have no file extension

File: client/package.json [hunk 35]
```diff
@@ -13,20 +13,31 @@
     "not ie < 9"
   ],
   "dependencies": {
+    "axios": "~0.19.0",
```
Comment (undead404) [inline client/package.json]: Remove tilde here, with zero versions maintainers doesn't provide any guarantees, let's stick to exact version.

File: client/src/routes/wind/LogRecord.js (hunk 26 above)
Comment (undead404) [reply client/src/routes/wind/LogRecord.js]: Much mess is to create a component called `Record` in a file named `LogRecord`. Please remove and rename files and components to match the current project structure.

File: client/src/routes/wind/RightPanel.js [hunk 36]
```diff
@@ -0,0 +1,159 @@
...
+      </>
+    );
+  }
+
+  getRotationDegree = dr => ({
+    transform: `rotate(${defineRotationDegree(dr)})`,
+  });
+
+  getAnimation = bv => ({
+    // eslint-disable-next-line css-modules/no-undef-class
```
Comment (undead404) [inline client/src/routes/wind/RightPanel.js]: What's the problem here?

File: client/src/routes/wind/WindChart.js [hunk 37]
```diff
@@ -0,0 +1,148 @@
...
+          : moment()
+              .startOf(period)
+              .toJSON(),
+      value: filter === period ? 'hour' : period,
+    });
+  };
+
+  parseLastSync = lastSync => {
+    const inactiveTime = 60;
+    const lsMoment = moment(lastSync);
```
Comment (undead404) [inline client/src/routes/wind/WindChart.js]: I don't insist this time, but short entity names are one of the most harmful things in programming

**Review by undead404 — CHANGES_REQUESTED** [review]

File: client/src/routes/wind/RightPanel.js (hunk 36 above)
Comment (Boto1) [reply client/src/routes/wind/RightPanel.js]: image
i dont know why, but it doenst see keyframes in css

File: client/src/routes/wind/WindChart/WindChart.js [hunk 38]
```diff
@@ -0,0 +1,147 @@
+/* eslint-disable react/jsx-handler-names */
```
Comment (undead404) [inline client/src/routes/wind/WindChart/WindChart.js]: Please remove this directive and refactor the component to have a handler only destructuring an event object and `fetchFilterData` receiving only period

File: package.json [hunk 39]
```diff
@@ -0,0 +1,5 @@
+{
+  "dependencies": {
+    "classnames": "^2.2.6"
```
Comment (undead404) [inline package.json]: Oops!

**Review by undead404 — CHANGES_REQUESTED** [review]

**Review by undead404 — APPROVED** [review]

File: server/domain/routers/wind.go [hunk 40]
```diff
@@ -0,0 +1,145 @@
...
+	w.Write(response)
+}
+
+func FindWindEvents(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+	keys := r.URL.Query()
+	from := keys.Get("from")
+	to := keys.Get("to")
+
+	wind := devices.Wind{}
+	windEvents, err := wind.FindManyEvents(from, to)
```
Comment (polluxx) [inline server/domain/routers/wind.go:98]: Added the same comment to Volodya PR:
 to prevent SQL injection please make time objects from that params and path them

File: server/go.mod [hunk 41]
```diff
@@ -3,6 +3,8 @@ module school-project-2019/server
 go 1.12
 
 require (
+	github.com/jasonlvhit/gocron v0.0.0-20190920201010-985d45da66c5
```
Comment (polluxx) [inline server/go.mod]: Remove it please if it's not used

File: generator/main.go [hunk 42]
```diff
@@ -0,0 +1,151 @@
+package main
```
Comment (undead404) [inline generator/main.go]: Hmmm, I'm not sure but having a file called `generator/main.go` generating only wind events seems weird. Could you refactor this to maybe have your wind-specific code in a separate file, and then only running it in main?

File: generator/main.go [hunk 43]
```diff
@@ -0,0 +1,151 @@
...
+
+func main() {
+	setInterval(GenerateWindEvents, 5)
+	http.ListenAndServe(":1234", nil)
+}
+
+func GenerateWindEvents() {
+	err := checkForStatus()
+	if err != nil {
+		fmt.Printf("Error: %v\n", err)
```
Comment (undead404) [inline generator/main.go]: `fmt` package has a separate method to print errors

File: generator/main.go [hunk 44]
```diff
@@ -0,0 +1,151 @@
...
+	}
+	fmt.Println("Wind event created")
+	req.Body.Close()
+}
+
+func getLastSpeed() float64 {
+	res, err := http.Get("http://localhost:8080/wind/event/last")
+	if err != nil {
+
+		return -1
```
Comment (undead404) [inline generator/main.go]: This is the place where even `-1` is a dangerous magic number. What does this speed mean? Server unaccessible? calculation failure?

File: server/domain/routers/wind.go (hunk 40 above)
Comment (undead404) [reply server/domain/routers/wind.go:98]: @Boto1 still not fixed?

File: server/go.mod (hunk 41 above)
Comment (undead404) [reply server/go.mod]: @Boto1 still not fixed?

**Review by undead404 — CHANGES_REQUESTED** [review]

File: generator/main.go (hunk 42 above)
Comment (undead404) [reply generator/main.go]: Please remove this file if it's unnecessary now

File: go.sum [hunk 45]
```diff
@@ -0,0 +1,147 @@
+cloud.google.com/go v0.26.0/go.mod h1:aQUYkXzVsufM+DwF1aE+0xfcU+56JwCaLick0ClmMTw=
```
Comment (undead404) [inline go.sum]: How did this appear in root dir?

**Review by undead404 — CHANGES_REQUESTED** [review]

**Review by undead404 — APPROVED** [review]
